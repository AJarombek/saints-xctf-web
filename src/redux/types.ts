/**
 * Types for the Redux store.
 * @author Andrew Jarombek
 * @since 7/23/2020
 */

export type RootState = {
    auth: AuthState,
    logs: LogsState
}

export type LogsState = {
    isFetching: boolean;
    didInvalidate: boolean;
    lastUpdated: number;
    items: Logs;
    feeds: LogFeeds;
    newComments: NewComments;
}

export type AuthState = {
    auth: Auth;
    user: Users;
}

export interface Meta {
    isFetching?: boolean;
    lastUpdated?: number;
    didInvalidate?: boolean;
    serverError?: string;
}

export type Auth = {
    isFetching?: boolean;
    lastUpdated?: number;
    signedInUser?: string;
    status?: string;
}

export type Users = {
    [key: string]: UserMeta;
}

export interface UserMeta extends User, Meta {}

export interface User {
    activation_code?: string;
    class_year?: number;
    deleted?: string;
    description?: string;
    email?: string;
    favorite_event?: string;
    first?: string;
    last?: string;
    last_signin?: string;
    location?: string;
    member_since?: string;
    password?: string;
    salt?: string;
    subscribed?: string;
    username?: string;
    week_start?: string;
}

enum LogType { RUN = "run", BIKE = "bike", SWIM = "swim", OTHER = "other" }
enum Metric { MILES = "miles", KILOMETERS = "kilometers", METERS = "meters" }

export type Log = {
    log_id: number;
    username: string;
    first: string;
    last: string;
    name: string;
    date: string;
    type: LogType;
    distance?: number;
    metric?: Metric;
    miles?: number;
    time?: string;
    pace?: string;
    feel: number;
    location?: string;
    description?: string;
    timeCreated?: string;
    comments?: Comment[];
}

export type Logs = {
    [key: string]: Record<string, Log>;
}

export type LogFeeds = {
    [key: string]: {
        filterBy: string;
        bucket: string;
        pages: LogFeedPages
    };
}

export type LogFeedPages = {
    [key: string]: LogFeedPage
}

export type LogFeedPage = {
    isFetching: boolean;
    lastUpdated: number;
    items: Log[];
    serverError: string;
}

export type Comment = {
    comment_id: number;
    username: string;
    first: string;
    last: string;
    log_id: number;
    time: string;
    content: string;
}

export type NewComments = {
    [key: string]: {
        isFetching: boolean;
        lastUpdated: number;
        created?: boolean;
        serverError?: string;
    }
}
