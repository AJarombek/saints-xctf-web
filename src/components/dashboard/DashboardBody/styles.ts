/**
 * JSS styles for the DashboardBody component.
 * @author Andrew Jarombek
 * @since 7/24/2020
 */

import Colors from "../../../styles/colors";

export default {
    dashboardBody: {
        display: 'grid',
        gridTemplateAreas: '"panel feed"',
        gridTemplateColumns: '1fr 2fr',
        minHeight: 'calc(100vh - 100px)',
        backgroundColor: Colors.lightBackground,
        margin: '100px 6% 0 6%',
    },
    sidePanel: {
        gridArea: 'panel',
        padding: '20px',
    },
    mainPanel: {
        gridArea: 'feed',
        padding: '20px',
    },
    '@media screen and (max-width: 1100px)': {
        dashboardBody: {
            margin: '100px 2% 0 2%',
        }
    },
    '@media screen and (max-width: 950px)': {
        dashboardBody: {
            margin: '100px 0 0 0',
        }
    }
};
