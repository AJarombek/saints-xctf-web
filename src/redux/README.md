### Overview

Configure Redux as the state management system for the application.

### Files

| Filename                 | Description                                                                |
|--------------------------|----------------------------------------------------------------------------|
| `modules`                | Redux modules containing Reducers, Actions, and Action Creators.           |
| `appState.js`            | Sample application state maintained by Redux.                              |
| `store.js`               | Configuration and creation of the Redux store.                             |
| `types.ts`               | Typescript types used throughout the application.                          |

### Resources

1) [React/Redux Example App #1](https://github.com/erikras/react-redux-universal-hot-example/tree/master/src/redux)
2) [React/Redux Example App #2](https://github.com/goopscoop/ga-react-tutorial/tree/6-reduxActionsAndReducers/src/redux)
3) [React/Redux Example App #3](https://github.com/gothinkster/react-redux-realworld-example-app/tree/master/src)
4) [Redux Ducks Spec](https://github.com/erikras/ducks-modular-redux)
5) [React Router/Redux Setup Steps](https://github.com/supasate/connected-react-router#usage)
6) [Redux Hot Reloading](https://redux.js.org/recipes/configuring-your-store/#hot-reloading)
7) [Redux Thunk](https://github.com/reduxjs/redux-thunk)
8) [Redux/React Official Real World Example](https://github.com/reduxjs/redux/tree/master/examples/real-world)
