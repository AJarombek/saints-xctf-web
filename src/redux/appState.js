/**
 * Sample Redux State for the entire application.
 * @author Andrew Jarombek
 * @since 5/3/2020
 */

const appState = {
  auth: {
    auth: {
      isFetching: false,
      lastUpdated: 1588530309,
      signedIn: true,
      status: "SUCCESS"
    },
    registration: {
      stage: 1,
      first: "Andy",
      last: "Jarombek",
      email: "andrew@jarombek.com",
      username: undefined,
      password: undefined,
      activationCode: undefined,
      teams: {
        saintsXCTFAlumni: {
          status: "accepted",
          groups: {
            alumni: {
              status: "accepted"
            }
          }
        }
      }
    },
    user: {
      andy: {
        isFetching: false,
        didInvalidate: false,
        lastUpdated: 1588530310,
        username: "andy",
        first: "Andrew",
        last: "Jarombek",
        password: "xxxxxx",
        description: "Test Description",
        memberSince: "2016-02-23 12:00:00",
        classYear: 2017,
        location: "New York, NY",
        favoriteEvent: "Cooldown Jog",
        activationCode: "ABCD1234",
        email: "andrew@jarombek.com",
        lastSignIn: "2020-05-03 12:00:00",
        subscribed: "Y",
        logs: []
      }
    },
  },
  profile: {
    tom: {
      isFetching: false,
      didInvalidate: false,
      lastUpdated: 1588530311
    },
    joe: {
      isFetching: false,
      didInvalidate: false,
      lastUpdated: 1588530312
    },
    ben: {
      isFetching: false,
      didInvalidate: false,
      lastUpdated: 1588530313
    }
  },
  group: {
    wmenstf: {
      isFetching: false,
      didInvalidate: false,
      lastUpdated: 1588530314,
      groupName: "wmenstf",
      groupTitle: "Women's Track & Field",
      description: "",
      weekStart: "monday",
      members: [],
      logs: []
    }
  },
  logs: {
    isFetching: false,
    didInvalidate: false,
    lastUpdated: 1588530315,
    items: {
      1: {
        log_id: 1,
        username: "andy",
        first: "Andrew",
        last: "Jarombek",
        name: "Reservoir Loop",
        location: "New York, NY",
        date: "2020-05-03",
        type: "run",
        distance: 6.37,
        metric: "miles",
        miles: 6.37,
        time: "00:45:37",
        pace: "00:07:10",
        feel: 6,
        description: "",
        timeCreated: "2020-05-03 12:00:00",
        comments: {
          1: {
            commentId: 1,
            username: "andy",
            first: "Andrew",
            last: "Jarombek",
            logId: 1,
            time: "2020-05-03 12:00:00",
            content: "test"
          }
        }
      }
    }
  }
};
