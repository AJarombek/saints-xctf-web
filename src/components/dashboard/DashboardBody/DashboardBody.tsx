/**
 * DashboardBody component which shows a paginated log feed and drawer of navigation links.
 * @author Andrew Jarombek
 * @since 7/24/2020
 */

import React, {useEffect, useMemo, useState} from 'react';
import {createUseStyles} from "react-jss";
import styles from "./styles";
import DashboardSidePanel from "../DashboardSidePanel/DashboardSidePanel";
import DashboardFeed from "../DashboardFeed/DashboardFeed";
import PaginationBar from "../../shared/PaginationBar/PaginationBar";
import {DeletedLogs, GroupMember, Log, LogFeeds, NewComments, NotificationsState, User} from "../../../redux/types";

interface IProps {
    getLogFeed: (filterBy: string, bucket: string, limit: number, offset: number) => void;
    postComment: (logId: number, username: string, first: string, last: string, content: string) => void;
    addComment: (logId: number, content: string, username: string, first: string, last: string,
        filterBy: string, bucket: string, page: number, index: number) => void;
    getGroupMemberships: (username: string) => void,
    getUserNotifications: (username: string) => void;
    deleteLog: (logId: number) => void;
    logFeeds: LogFeeds;
    newComments: NewComments;
    user: User;
    groupMemberships: GroupMember[];
    notificationInfo: NotificationsState;
    deletedLogs: DeletedLogs;
}

const useStyles = createUseStyles(styles);

const DashboardBody: React.FunctionComponent<IProps> = ({
    getLogFeed,
    postComment,
    addComment,
    getGroupMemberships,
    getUserNotifications,
    deleteLog,
    logFeeds,
    newComments,
    user,
    groupMemberships,
    notificationInfo,
    deletedLogs
}) => {
    const classes = useStyles();

    const [page, setPage] = useState(1);

    const filterBy = "all";
    const bucket = "all";

    useEffect(() => {
        getLogFeed(filterBy, bucket, 10, 10 * (page - 1));
    }, [filterBy, bucket, page]);

    useEffect(() => {
        if (user) {
            if (!groupMemberships?.length) {
                getGroupMemberships(user.username);
            }

            getUserNotifications(user.username);
        }
    }, [user]);

    const totalPages: number = useMemo(() => {
        return logFeeds[`${filterBy}-${bucket}`]?.pages[page]?.pages ?? 0
    }, [logFeeds, page]);

    return (
        <div className={classes.dashboardBody}>
            <div className={classes.sidePanel}>
                <DashboardSidePanel
                    user={user}
                    groupMemberships={groupMemberships}
                    notificationInfo={notificationInfo}
                />
            </div>
            <div className={classes.mainPanel}>
                <DashboardFeed
                    logFeeds={logFeeds}
                    getLogFeed={getLogFeed}
                    postComment={postComment}
                    addComment={addComment}
                    deleteLog={deleteLog}
                    page={page}
                    newComments={newComments}
                    deletedLogs={deletedLogs}
                    user={user}
                    filterBy={filterBy}
                    bucket={bucket}
                />
                <PaginationBar
                    page={page}
                    totalPages={totalPages}
                    onChangePage={(page) => {
                        setPage(page);
                        window.scrollTo(0, 0);
                    }}
                />
            </div>
        </div>
    );
};

export default DashboardBody;
