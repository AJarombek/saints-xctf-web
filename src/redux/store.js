/**
 * Build and configure the redux store, which holds the application state and handles
 * all state updates.
 * @author Andrew Jarombek
 * @since 5/1/2020
 */

import { createBrowserHistory } from 'history';
import { createStore, applyMiddleware } from 'redux';
import { routerMiddleware } from 'connected-react-router';
import createLogger from 'redux-logger';
import reducer from './modules/reducers';

const history = createBrowserHistory();

const configureStore = (data) => {
  const loggerMiddleware = createLogger();
  const middleware = [loggerMiddleware, routerMiddleware(history)];

  const createStoreWithMiddleware = applyMiddleware(...middleware)(createStore);
  return createStoreWithMiddleware(reducer, data);
};

export { history };
export default configureStore;
