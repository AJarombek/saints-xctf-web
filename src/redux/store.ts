/**
 * Build and configure the redux store, which holds the application state and handles
 * all state updates.
 * @author Andrew Jarombek
 * @since 5/1/2020
 */

import { createBrowserHistory } from 'history';
import { createStore, applyMiddleware, Store, Action } from 'redux';
import { routerMiddleware } from 'connected-react-router';
import { createLogger } from 'redux-logger';
import thunkMiddleware, { ThunkAction } from 'redux-thunk';
import reducer from './modules/reducers';
import { RootState } from './types';

export const history = createBrowserHistory();

export default function configureStore(): Store {
  const loggerMiddleware = createLogger();
  const middleware = [routerMiddleware(history), loggerMiddleware, thunkMiddleware];

  const createStoreWithMiddleware = applyMiddleware(...middleware)(createStore);
  const store = createStoreWithMiddleware(reducer(history));

  if (process.env.NODE_ENV !== 'production' && module.hot) {
    module.hot.accept('./modules/reducers', () => store.replaceReducer(reducer));
  }

  return store;
}

export type AppThunk<ReturnType = void, State = RootState> = ThunkAction<ReturnType, State, unknown, Action<string>>;
