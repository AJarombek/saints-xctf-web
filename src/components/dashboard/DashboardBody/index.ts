import DashboardBody from './DashboardBody';

export default DashboardBody;
