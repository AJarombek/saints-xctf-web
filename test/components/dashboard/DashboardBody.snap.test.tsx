/**
 * Snapshot test for the DashboardBody component.
 * @author Andrew Jarombek
 * @since 3/4/2021
 */

import React from 'react';
import renderer from 'react-test-renderer';
import { Store } from 'redux';
import { emptyStore } from '../../test-utils/storeMocks';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { andy } from '../../test-utils/userMocks';
import DashboardBody from '../../../src/components/dashboard/DashboardBody';

const mockStore = configureStore([thunk]);

describe('DashboardBody Snapshot Tests', () => {
  let store: Store;

  beforeEach(() => {
    store = mockStore(emptyStore);
  });

  it('renders correctly', () => {
    const tree = renderer
      .create(
        <MemoryRouter>
          <Provider store={store}>
            <DashboardBody user={andy} />
          </Provider>
        </MemoryRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
