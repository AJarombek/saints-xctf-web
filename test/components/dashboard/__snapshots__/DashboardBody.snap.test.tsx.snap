// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DashboardBody Snapshot Tests renders correctly 1`] = `
<div
  className="dashboardBody-0-1-1"
>
  <div
    className="sidePanel-0-1-2"
  >
    <div
      className="dashboardSidePanel-0-1-4"
      id="dashboardSidePanel"
    >
      <div
        className="accordion-0-1-17 accordion"
        onClick={[Function]}
      >
        <div
          className="header-0-1-18"
          onClick={[Function]}
        >
          <p
            className="icon-0-1-19"
          >
            <p>
              
            </p>
          </p>
          <div
            className="title-0-1-20"
          >
            Profile
          </div>
          <p
            className="arrow-0-1-21 expandIcon"
          >
            E
          </p>
        </div>
      </div>
      <div
        className="accordion-0-1-17 accordion"
        onClick={[Function]}
      >
        <div
          className="header-0-1-18"
          onClick={[Function]}
        >
          <p
            className="icon-0-1-19"
          >
            <p>
              P
            </p>
          </p>
          <div
            className="title-0-1-20"
          >
            Create New Log
          </div>
          <p
            className="arrow-0-1-21 expandIcon"
          >
            E
          </p>
        </div>
      </div>
      <div
        className="accordion-0-1-17 accordion"
        id="groupsAccordion"
      >
        <div
          className="header-0-1-18"
          onClick={[Function]}
        >
          <p
            className="icon-0-1-19"
          >
            <p>
              
            </p>
          </p>
          <div
            className="title-0-1-20"
          >
            Teams & Groups
          </div>
          <p
            className="arrow-0-1-21 expandIcon"
          >
            B
          </p>
        </div>
        <div
          className="body-0-1-22"
        />
      </div>
      <div
        className="accordion-0-1-17 accordion"
      >
        <div
          className="header-0-1-18"
          onClick={[Function]}
        >
          <p
            className="icon-0-1-19"
          >
            <div
              className="notificationCount-0-1-10 noNotifications-0-1-12"
            >
              <div
                className="aj-notification-circle container-2-2-1"
              >
                <p
                  className="count-2-2-2"
                >
                  0
                </p>
              </div>
            </div>
          </p>
          <div
            className="title-0-1-20"
          >
            Notifications
          </div>
          <p
            className="arrow-0-1-21 expandIcon"
          >
            C
          </p>
        </div>
      </div>
    </div>
  </div>
  <div
    className="mainPanel-0-1-3"
  >
    <div
      className="logFeed-0-1-23"
      id="logFeed"
    >
      <div
        className="loading-0-1-24"
      >
        <div
          className="aj-loading-dots container-2-2-3"
        >
          <div
            className="aj-loading-circle circle-2-2-4 first-2-2-5"
          />
          <div
            className="aj-loading-circle circle-2-2-4 second-2-2-6"
          />
          <div
            className="aj-loading-circle circle-2-2-4 third-2-2-7"
          />
        </div>
      </div>
    </div>
    <div
      className="paginationBar-0-1-25"
      id="paginationBar"
    >
      <div
        className="page-0-1-26 currentPage-0-1-28"
      >
        1
      </div>
    </div>
  </div>
</div>
`;
